README
XBMC Vu+ PVR Add-on
------------------------------------------

THIS IS A PRELIMINARY README AND IS SUBJECT TO CHANGE!!!

Written by:                  Joerg Dembski

Project's homepage:          

Latest version available at: 

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

------------------------------------------

This is a PVR Add-on for XBMC to add vu+ DVRs as a TV/PVR Backend to XBMC.

It want to add support for Live TV watching, replaying of Recordings, programming Timers and
EPG TV Guide to use over the Network.

For parsing the XML output from the VU+ webinterface it uses the xmlParser library developed by Frank Vanden Berghen (http://www.applied-mathematics.net/tools/xmlParser.html). The build scripts are based on Alexander Pipelka's xbmc-addon-xvdr. Many thanks for this!    

---------------------------------------------
ADDON COMPILATION & INSTALLATION INSTRUCTIONS

IMPORTANT:
Please disable *all* PVR addons *before* installing the VU+ addon!

- Clone the GIT repository
- cd xbmc-addon-vuplus
- sh autogen.sh
- ./configure --prefix=/usr/lib/xbmc (or $HOME/.xbmc)
- make
- make install

This will install the plugin into your /usr/lib/xbmc/addons or $HOME/.xbmc/addons directory


ChangeLog
0.2.9:
- change: remove curl dependancy (should also fix compiling unter IOS)

0.2.8:
- add: option to only fetch one TV bouquet which can be specified via the addon settings

0.2.7:
- fix: fix incorrect escape character for & (use '&amp;' instead of '&quot;'). Thx to 'hani' for pointing this out.

0.2.6:
- cosmetic: remove unnecessary '/' in recording-stream url
- cosmetic: inprove log output 
- change: get the proper device info from the reveiver box instead of just setting dummy values
- change: change the buildzip.bat to include version string in the name of the zip-file
- change: introduce a version string for the channeldata xml file so that we can invalidate old channeldata files if necessary

0.2.5:
- fix: cleanup the service reference in the streaming url

0.2.4
- fix: put username & password into the streaming url

0.2.3
- fix: put username & password into recording-url

0.2.2:
- fix: escape xml entities in the stream-url before saving to channeldata.xml
- fix: fetch the version string from configure.in not the ChangeLog (thanks to 'trans')
- fix: rename changelog in addon-directory to changelog.txt

0.2.1:
- fix: encode the stream-url properly. Please remove your channeldata.xml file after updating
- change: add proper version string to the addon.xml generated by buildzip.bat (thanks to 'trans')

0.2.0:
- change: Complete rework of timer handling. Now it will only show notifications for changed items
- fix: Update / Add / Delete Timers. Note that the timer names in the XBMC are sometimes not updated properly (in the dialog it is OK)
- fix: Instant Recordings do now work again
- fix: timer status is now shown correctly (completed, recording, disabled)
- fix: enabling / disabling timers do now work
- change: removed "show completed timers" option (they are now always be shown)
- add: added buildzip.bat & xbmc-addon-vuplus.sln (thanks to 'trans')
- fix: Channel numbering / incorrect picons due to the change introduced in 0.1.15

0.1.15
- fix: when using the channeldata-store option not all channel groups (bouquets) were loaded due to a nulled counter variable

0.1.14
- change: change .gitignores and copy current changelog into the addon folder. This will enable the user to view the changelog in the addon settings window

0.1.13 
- add: option to set deepstandby powerstate on the DVB box on PVR-addon shutdown
- fix: error in settings.xml, causing crashes on linux

0.1.12 
- add: support to specify the folder for storing recordings

0.1.11 
- fix: some unicode string errors in german language file
- change: add categories to addon settings

0.1.10 
- add: support for using only the recording path currently active on the DVB box 

0.1.9 
- change: add addon icon (copyied from the openelec guys, thanks!)
- fix: win32 release target (thanks to 'trans' for patch)

0.1.8 
- change: fetch the recording locations and use these locations when querying the recordings
- fix: Stop update thread when destroying the PVR addon - fixes shutdown issues

0.1.7 
- fix: Copy & Paste error                        

0.1.6 
- fix: Change the logic to determine the picon file name from the service reference.

0.1.5 
- fix: Fix access violation errors on windows (thanks to 'trans' for finding the correct compiler / linker settings)
- fix: do not return out of scope local variable for the streaming URL
- fix: remove the last '_' character from the filename (if applicable). Should finally fix the picon path errors.

0.1.4 
- fix: limit length of the service reference part of the Icon-Path to 30 characters. 
- fix: Escape "&", "<" and ">" characters in XML channeldata file.
- add: support to disable all bouquet or channel updates 

0.1.3 
- fix: order of timerlist cleanup and timerlist updates 

0.1.2 
- add: support for storing channel data into a file

0.1.1 
- fix: URLEncode the service reference when fetching EPG

0.1.0 
- Iniital Version


src/vuplus/VuData.h
#pragma once 

#include "../platform/util/StdString.h"
#include "xmlParser.h"
#include "client.h"
#include "../platform/threads/threads.h"
    
struct VuWebResponse {
  char *response;
  int iSize;
};

#define CHANNELDATAVERSION  2

typedef enum VU_UPDATE_STATE
{
    VU_UPDATE_STATE_NONE,
    VU_UPDATE_STATE_FOUND,
    VU_UPDATE_STATE_UPDATED,
    VU_UPDATE_STATE_NEW
} VU_UPDATE_STATE;

struct VuChannelGroup {
  std::string strServiceReference;
  std::string strGroupName;
  int iGroupState;

  VuChannelGroup() 
  { 
    iGroupState = VU_UPDATE_STATE_NEW;
  }

  bool operator==(const VuChannelGroup &right) const
  {
    return (! strServiceReference.compare(right.strServiceReference)) && (! strGroupName.compare(right.strGroupName));
  }

};

struct VuChannel
{
  bool bRadio;
  int iUniqueId;
  int iChannelNumber;
  std::string strGroupName;
  std::string strChannelName;
  std::string strServiceReference;
  std::string strStreamURL;
  std::string strIconPath;
  int iChannelState;

  VuChannel()
  {
    iChannelState = VU_UPDATE_STATE_NEW;
  }
  
  bool operator==(const VuChannel &right) const
  {
    bool bChanged = true;
    bChanged = bChanged && (bRadio == right.bRadio); 
    bChanged = bChanged && (iUniqueId == right.iUniqueId); 
    bChanged = bChanged && (iChannelNumber == right.iChannelNumber); 
    bChanged = bChanged && (! strGroupName.compare(right.strGroupName));
    bChanged = bChanged && (! strChannelName.compare(right.strChannelName));
    bChanged = bChanged && (! strServiceReference.compare(right.strServiceReference));
    bChanged = bChanged && (! strStreamURL.compare(right.strStreamURL));
    bChanged = bChanged && (! strIconPath.compare(right.strIconPath));

    return bChanged;
  }

};

struct VuEPGEntry 
{
  int iEventId;
  std::string strServiceReference;
  std::string strTitle;
  int iChannelId;
  time_t startTime;
  time_t endTime;
  std::string strPlotOutline;
  std::string strPlot;
};

struct VuTimer
{
  std::string strTitle;
  std::string strPlot;
  int iChannelId;
  time_t startTime;
  time_t endTime;
  bool bRepeating; 
  int iWeekdays;
  int iEpgID;
  PVR_TIMER_STATE state; 
  int iUpdateState;
  unsigned int iClientIndex;

  VuTimer()
  {
    iUpdateState = VU_UPDATE_STATE_NEW;
  }
  
  bool like(const VuTimer &right) const
  {
    bool bChanged = true;
    bChanged = bChanged && (startTime == right.startTime); 
    bChanged = bChanged && (endTime == right.endTime); 
    bChanged = bChanged && (iChannelId == right.iChannelId); 
    bChanged = bChanged && (bRepeating == right.bRepeating); 
    bChanged = bChanged && (iWeekdays == right.iWeekdays); 
    bChanged = bChanged && (iEpgID == right.iEpgID); 

    return bChanged;
  }
  
  bool operator==(const VuTimer &right) const
  {
    bool bChanged = true;
    bChanged = bChanged && (startTime == right.startTime); 
    bChanged = bChanged && (endTime == right.endTime); 
    bChanged = bChanged && (iChannelId == right.iChannelId); 
    bChanged = bChanged && (bRepeating == right.bRepeating); 
    bChanged = bChanged && (iWeekdays == right.iWeekdays); 
    bChanged = bChanged && (iEpgID == right.iEpgID); 
    bChanged = bChanged && (state == right.state); 
    bChanged = bChanged && (! strTitle.compare(right.strTitle));
    bChanged = bChanged && (! strPlot.compare(right.strPlot));

    return bChanged;
  }
};

struct VuRecording
{
  std::string strRecordingId;
  time_t startTime;
  int iDuration;
  std::string strTitle;
  std::string strStreamURL;
  std::string strPlot;
  std::string strPlotOutline;
  std::string strChannelName;
};
 
class Vu  : public PLATFORM::CThread
{
private:

  // members
  std::string m_strEnigmaVersion;
  std::string m_strImageVersion;
  std::string m_strWebIfVersion;
  bool  m_bIsConnected;
  std::string m_strServerName;
  std::string m_strURL;
  int m_iNumRecordings;
  int m_iNumChannelGroups;
  int m_iCurrentChannel;
  unsigned int m_iUpdateTimer;
  std::vector<VuChannel> m_channels;
  std::vector<VuTimer> m_timers;
  std::vector<VuRecording> m_recordings;
  std::vector<VuChannelGroup> m_groups;
  std::vector<std::string> m_locations;

  bool m_bInitial;
  unsigned int m_iClientIndexCounter;

  PLATFORM::CMutex m_mutex;
  PLATFORM::CCondition<bool> m_started;
 

  // functions

  void StoreChannelData();
  void LoadChannelData();
  CStdString GetHttpXML(CStdString& url);
  int GetChannelNumber(CStdString strServiceReference);
  CStdString URLEncodeInline(const CStdString& strData);
  bool SendSimpleCommand(const CStdString& strCommandURL, CStdString& strResult, bool bIgnoreResult = false);
  static int VuWebResponseCallback(void *contents, int iLength, int iSize, void *memPtr); 
  CStdString GetGroupServiceReference(CStdString strGroupName);
  bool LoadChannels(CStdString strServerReference, CStdString strGroupName);
  bool LoadChannels();
  bool LoadChannelGroups();
  bool LoadLocations();
  std::vector<VuTimer> LoadTimers();
  void TimerUpdates();

  // helper functions
  static bool GetInt(XMLNode xRootNode, const char* strTag, int& iIntValue);
  static bool GetBoolean(XMLNode xRootNode, const char* strTag, bool& bBoolValue);
  static bool GetString(XMLNode xRootNode, const char* strTag, CStdString& strStringValue);
  static long TimeStringToSeconds(const CStdString &timeString);
  static int SplitString(const CStdString& input, const CStdString& delimiter, CStdStringArray &results, unsigned int iMaxStrings = 0);
  bool CheckForGroupUpdate();
  bool CheckForChannelUpdate();
  std::string& Escape(std::string &s, std::string from, std::string to);
  bool GetDeviceInfo();


protected:
  virtual void *Process(void);

public:
  Vu(void);
  ~Vu();

  const char * GetServerName();
  bool IsConnected(); 
  int GetChannelsAmount(void);
  PVR_ERROR GetChannels(PVR_HANDLE handle, bool bRadio);
  PVR_ERROR GetEPGForChannel(PVR_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd);
  int GetCurrentClientChannel(void);
  int GetTimersAmount(void);
  PVR_ERROR GetTimers(PVR_HANDLE handle);
  PVR_ERROR AddTimer(const PVR_TIMER &timer);
  PVR_ERROR UpdateTimer(const PVR_TIMER &timer);
  PVR_ERROR DeleteTimer(const PVR_TIMER &timer);
  bool GetRecordingFromLocation(PVR_HANDLE handle, CStdString strRecordingFolder);
  unsigned int GetRecordingsAmount();
  PVR_ERROR    GetRecordings(PVR_HANDLE handle);
  PVR_ERROR    DeleteRecording(const PVR_RECORDING &recinfo);
  unsigned int GetNumChannelGroups(void);
  PVR_ERROR    GetChannelGroups(PVR_HANDLE handle);
  PVR_ERROR    GetChannelGroupMembers(PVR_HANDLE handle, const PVR_CHANNEL_GROUP &group);
  const char* GetLiveStreamURL(const PVR_CHANNEL &channelinfo);
  bool OpenLiveStream(const PVR_CHANNEL &channelinfo);
  void CloseLiveStream();
  void SendPowerstate();
  bool SwitchChannel(const PVR_CHANNEL &channel);
  bool Open();
  void Action();
};



src/vuplus/client.h
#pragma once
/*
 *      Copyright (C) 2005-2011 Team XBMC
 *      http://www.xbmc.org
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XBMC; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */

#include "../../addons/library.xbmc.addon/libXBMC_addon.h"
#include "../../addons/library.xbmc.pvr/libXBMC_pvr.h"

#define DEFAULT_HOST             "127.0.0.1"
#define DEFAULT_CONNECT_TIMEOUT  30
#define DEFAULT_STREAM_PORT      8001 
#define DEFAULT_WEB_PORT         80
#define DEFAULT_UPDATE_INTERVAL  2

extern bool                      m_bCreated;
extern std::string               g_strHostname;
extern int                       g_iPortStream;
extern int                       g_iPortWeb;
extern std::string               g_strUsername;
extern std::string               g_strPassword;
extern std::string               g_strIconPath;
extern std::string               g_strRecordingPath;
extern int 			 g_iUpdateInterval;
extern int                       g_iClientId;
extern unsigned int              g_iPacketSequence;
extern bool                      g_bShowTimerNotifications;
extern bool			 g_bZap;
extern bool                      g_bAutomaticTimerlistCleanup;
extern bool                      g_bCheckForGroupUpdates;
extern bool                      g_bCheckForChannelUpdates;
extern bool                      g_bOnlyCurrentLocation;
extern bool			 g_bSetPowerstate;
extern bool			 g_bOnlyOneGroup;
extern std::string               g_strOneGroup;
extern std::string               g_szUserPath;
extern std::string               g_szClientPath;
extern std::string               g_strChannelDataPath;
extern ADDON::CHelper_libXBMC_addon *   XBMC;
extern CHelper_libXBMC_pvr *     PVR;


src/vuplus/client.cpp
/*
 *      Copyright (C) 2005-2011 Team XBMC
 *      http://xbmc.org
 *
 *  This Program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  This Program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with XBMC; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  http://www.gnu.org/copyleft/gpl.html
 *
 */

#include "client.h"
#include "xbmc_pvr_dll.h"
#include <stdlib.h>
#include "VuData.h"

using namespace std;
using namespace ADDON;

bool         m_bCreated  = false;
ADDON_STATUS m_CurStatus = ADDON_STATUS_UNKNOWN;
int          g_iClientId = -1;

/* User adjustable settings are saved here.
 * Default values are defined inside client.h
 * and exported to the other source files.
 */
std::string g_strHostname             = DEFAULT_HOST;
int         g_iConnectTimeout         = DEFAULT_CONNECT_TIMEOUT;
int         g_iPortStream             = DEFAULT_STREAM_PORT;
int         g_iPortWeb                = DEFAULT_WEB_PORT;
int         g_iUpdateInterval         = DEFAULT_UPDATE_INTERVAL;
std::string g_strUsername             = "";
std::string g_strRecordingPath        = "";
std::string g_strPassword             = "";
std::string g_szUserPath              = "";
std::string g_strIconPath             = "";
bool        g_bAutomaticTimerlistCleanup = false;
bool        g_bZap                    = false;
bool        g_bCheckForGroupUpdates   = true;
bool        g_bCheckForChannelUpdates = true;
bool        g_bOnlyCurrentLocation    = false;
bool        g_bSetPowerstate          = false;
bool        g_bOnlyOneGroup           = false;
std::string g_strOneGroup             = "";
std::string g_szClientPath            = "";
std::string g_strChannelDataPath      = "/tmp/";

CHelper_libXBMC_addon *XBMC           = NULL;
CHelper_libXBMC_pvr   *PVR            = NULL;
Vu                *VuData             = NULL;

extern "C" {

void ADDON_ReadSettings(void)
{
  /* read setting "host" from settings.xml */
  char * buffer;
  buffer = (char*) malloc (1024);
  buffer[0] = 0; /* Set the end of string */

  if (XBMC->GetSetting("host", buffer))
    g_strHostname = buffer;
  else
    g_strHostname = DEFAULT_HOST;
  buffer[0] = 0; /* Set the end of string */

  /* read setting "user" from settings.xml */
  if (XBMC->GetSetting("user", buffer))
    g_strUsername = buffer;
  else
    g_strUsername = "";
  buffer[0] = 0; /* Set the end of string */
  
  /* read setting "recordingpath" from settings.xml */
  if (XBMC->GetSetting("recordingpath", buffer))
    g_strRecordingPath = buffer;
  else
    g_strRecordingPath = "";
  buffer[0] = 0; /* Set the end of string */

  /* read setting "pass" from settings.xml */
  if (XBMC->GetSetting("pass", buffer))
    g_strPassword = buffer;
  else
    g_strPassword = "";
  
  /* read setting "streamport" from settings.xml */
  if (!XBMC->GetSetting("streamport", &g_iPortStream))
    g_iPortStream = DEFAULT_STREAM_PORT;
  
  /* read setting "webport" from settings.xml */
  if (!XBMC->GetSetting("webport", &g_iPortWeb))
    g_iPortWeb = DEFAULT_WEB_PORT;
  
  /* read setting "onlycurrent" from settings.xml */
  if (!XBMC->GetSetting("onlycurrent", &g_bOnlyCurrentLocation))
    g_bOnlyCurrentLocation = false;
  
  /* read setting "setpowerstate" from settings.xml */
  if (!XBMC->GetSetting("setpowerstate", &g_bSetPowerstate))
    g_bSetPowerstate = false;
  
  /* read setting "checkgroups" from settings.xml */
  if (!XBMC->GetSetting("checkgroups", &g_bCheckForGroupUpdates))
    g_bCheckForGroupUpdates = true;
  
  /* read setting "showcompleted" from settings.xml */
  if (!XBMC->GetSetting("checkchannels", &g_bCheckForChannelUpdates))
    g_bCheckForChannelUpdates = true;
  
  /* read setting "zap" from settings.xml */
  if (!XBMC->GetSetting("zap", &g_bZap))
    g_bZap = false;

  /* read setting "onlyonegroup" from settings.xml */
  if (!XBMC->GetSetting("onlyonegroup", &g_bOnlyOneGroup))
    g_bOnlyOneGroup = false;
  
  /* read setting "onegroup" from settings.xml */
  if (XBMC->GetSetting("onegroup", buffer))
    g_strOneGroup = buffer;
  else
    g_strOneGroup = "";

  /* read setting "timerlistcleanup" from settings.xml */
  if (!XBMC->GetSetting("timerlistcleanup", &g_bAutomaticTimerlistCleanup))
    g_bAutomaticTimerlistCleanup = false;

  /* read setting "updateint" from settings.xml */
  if (!XBMC->GetSetting("updateint", &g_iUpdateInterval))
    g_iConnectTimeout = DEFAULT_UPDATE_INTERVAL;

  /* read setting "iconpath" from settings.xml */
  if (XBMC->GetSetting("iconpath", buffer))
    g_strIconPath = buffer;
  else
    g_strIconPath = "";
  
  /* read setting "channeldatapath" from settings.xml */
  if (XBMC->GetSetting("channeldatapath", buffer))
    g_strChannelDataPath = buffer;
  else
    g_strChannelDataPath = "/tmp/";
  
  free (buffer);
}

ADDON_STATUS ADDON_Create(void* hdl, void* props)
{
  if (!hdl || !props)
    return ADDON_STATUS_UNKNOWN;

  PVR_PROPERTIES* pvrprops = (PVR_PROPERTIES*)props;

  XBMC = new CHelper_libXBMC_addon;
  if (!XBMC->RegisterMe(hdl))
  {
    delete XBMC;
    XBMC = NULL;
    return ADDON_STATUS_UNKNOWN;
  }

  PVR = new CHelper_libXBMC_pvr;
  if (!PVR->RegisterMe(hdl))
  {
    delete PVR;
    delete XBMC;
    PVR = NULL;
    XBMC = NULL;
    return ADDON_STATUS_UNKNOWN;
  }

  XBMC->Log(LOG_DEBUG, "%s - Creating VU+ PVR-Client", __FUNCTION__);

  m_CurStatus     = ADDON_STATUS_UNKNOWN;
  g_iClientId     = pvrprops->iClientId;
  g_szUserPath   = pvrprops->strUserPath;
  g_szClientPath  = pvrprops->strClientPath;

  ADDON_ReadSettings();

  VuData = new Vu;
  if (!VuData->Open()) 
  {
    delete VuData;
    delete PVR;
    delete XBMC;
    VuData = NULL;
    PVR = NULL;
    XBMC = NULL;
    m_CurStatus = ADDON_STATUS_LOST_CONNECTION;
    return m_CurStatus;
  }

  m_CurStatus = ADDON_STATUS_OK;
  m_bCreated = true;
  return m_CurStatus;
}

ADDON_STATUS ADDON_GetStatus()
{
  /* check whether we're still connected */
  if (m_CurStatus == ADDON_STATUS_OK && !VuData->IsConnected())
    m_CurStatus = ADDON_STATUS_LOST_CONNECTION;

  return m_CurStatus;
}

void ADDON_Destroy()
{
  if (m_bCreated)
  {
    m_bCreated = false;
  }

  if (VuData)
  {
    VuData->SendPowerstate();
  }
  

  if (PVR)
  {
    delete PVR;
    PVR = NULL;
  }

  if (XBMC)
  {
    delete XBMC;
    XBMC = NULL;
  }

  if (VuData)
  {
    delete VuData;
    VuData = NULL;
  }

  m_CurStatus = ADDON_STATUS_UNKNOWN;
}

bool ADDON_HasSettings()
{
  return true;
}

unsigned int ADDON_GetSettings(ADDON_StructSetting ***sSet)
{
  return 0;
}

ADDON_STATUS ADDON_SetSetting(const char *settingName, const void *settingValue)
{
  string str = settingName;
  if (str == "host")
  {
    string tmp_sHostname;
    XBMC->Log(LOG_INFO, "%s - Changed Setting 'host' from %s to %s", __FUNCTION__, g_strHostname.c_str(), (const char*) settingValue);
    tmp_sHostname = g_strHostname;
    g_strHostname = (const char*) settingValue;
    if (tmp_sHostname != g_strHostname)
      return ADDON_STATUS_NEED_RESTART;
  }
  else if (str == "user")
  {
    string tmp_sUsername = g_strUsername;
    g_strUsername = (const char*) settingValue;
    if (tmp_sUsername != g_strUsername)
    {
      XBMC->Log(LOG_INFO, "%s - Changed Setting 'user'", __FUNCTION__);
      return ADDON_STATUS_NEED_RESTART;
    }
  }
  else if (str == "pass")
  {
    string tmp_sPassword = g_strPassword;
    g_strPassword = (const char*) settingValue;
    if (tmp_sPassword != g_strPassword)
    {
      XBMC->Log(LOG_INFO, "%s - Changed Setting 'pass'", __FUNCTION__);
      return ADDON_STATUS_NEED_RESTART;
    }
  }
  else if (str == "streamport")
  {
    int iNewValue = *(int*) settingValue + 1;
    if (g_iPortStream != iNewValue)
    {
      XBMC->Log(LOG_INFO, "%s - Changed Setting 'streamport' from %u to %u", __FUNCTION__, g_iPortStream, iNewValue);
      g_iPortStream = iNewValue;
      return ADDON_STATUS_OK;
    }
  }
  else if (str == "webport")
  {
    int iNewValue = *(int*) settingValue + 1;
    if (g_iPortWeb != iNewValue)
    {
      XBMC->Log(LOG_INFO, "%s - Changed Setting 'webport' from %u to %u", __FUNCTION__, g_iPortWeb, iNewValue);
      g_iPortWeb = iNewValue;
      return ADDON_STATUS_OK;
    }
  }
  return ADDON_STATUS_OK;
}

void ADDON_Stop()
{
}

void ADDON_FreeSettings()
{
}

/***********************************************************
 * PVR Client AddOn specific public library functions
 ***********************************************************/

PVR_ERROR GetAddonCapabilities(PVR_ADDON_CAPABILITIES* pCapabilities)
{
  pCapabilities->bSupportsChannelSettings = false;
  pCapabilities->bSupportsTimeshift       = false;
  pCapabilities->bSupportsEPG             = true;
  pCapabilities->bSupportsTV              = true;
  pCapabilities->bSupportsRadio           = true;
  pCapabilities->bSupportsRecordings      = true;
  pCapabilities->bSupportsTimers          = true;
  pCapabilities->bSupportsChannelGroups   = true;
  pCapabilities->bSupportsChannelScan     = false;
  pCapabilities->bHandlesInputStream      = true;
  pCapabilities->bHandlesDemuxing         = false;

  return PVR_ERROR_NO_ERROR;
}

const char *GetBackendName(void)
{
  static const char *strBackendName = VuData ? VuData->GetServerName() : "unknown";
  return strBackendName;
}

const char *GetBackendVersion(void)
{
  static const char *strBackendVersion = "UNKNOWN";
  return strBackendVersion;
}

const char *GetConnectionString(void)
{
  CStdString strConnectionString;
  if (VuData)
    strConnectionString.Format("%s%s", g_strHostname.c_str(), VuData->IsConnected() ? "" : " (Not connected!)");
  else
    strConnectionString.Format("%s (addon error!)", g_strHostname.c_str());
  return strConnectionString.c_str();
}

PVR_ERROR GetDriveSpace(long long *iTotal, long long *iUsed)
{
  return PVR_ERROR_SERVER_ERROR;
}

PVR_ERROR GetEPGForChannel(PVR_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetEPGForChannel(handle, channel, iStart, iEnd);
}

int GetChannelsAmount(void)
{
  if (!VuData || !VuData->IsConnected())
    return 0;

  return VuData->GetChannelsAmount();
}

PVR_ERROR GetChannels(PVR_HANDLE handle, bool bRadio)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetChannels(handle, bRadio);
}

int GetRecordingsAmount(void)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetRecordingsAmount();
}

PVR_ERROR GetRecordings(PVR_HANDLE handle)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetRecordings(handle);
}

PVR_ERROR DeleteRecording(const PVR_RECORDING &recording)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->DeleteRecording(recording);
}

PVR_ERROR RenameRecording(const PVR_RECORDING &recording)
{
  return PVR_ERROR_NOT_IMPLEMENTED;
}

int GetTimersAmount(void)
{
  if (!VuData || !VuData->IsConnected())
    return 0;

  return VuData->GetTimersAmount();
}

PVR_ERROR GetTimers(PVR_HANDLE handle)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetTimers(handle);
}

PVR_ERROR AddTimer(const PVR_TIMER &timer)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->AddTimer(timer);
}

PVR_ERROR DeleteTimer(const PVR_TIMER &timer, bool bForceDelete)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->DeleteTimer(timer);
}

PVR_ERROR UpdateTimer(const PVR_TIMER &timer)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->UpdateTimer(timer);
}

int GetCurrentClientChannel(void)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetCurrentClientChannel();
}

bool SwitchChannel(const PVR_CHANNEL &channel)
{
  if (!VuData || !VuData->IsConnected())
    return false;

  return VuData->SwitchChannel(channel);
}

int GetChannelGroupsAmount(void)
{
  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetNumChannelGroups();
}

PVR_ERROR GetChannelGroups(PVR_HANDLE handle, bool bRadio)
{
  if (bRadio)
    return PVR_ERROR_NO_ERROR;

  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetChannelGroups(handle);
}

PVR_ERROR GetChannelGroupMembers(PVR_HANDLE handle, const PVR_CHANNEL_GROUP &group)
{
  if (group.bIsRadio)
    return PVR_ERROR_NO_ERROR;

  if (!VuData || !VuData->IsConnected())
    return PVR_ERROR_SERVER_ERROR;

  return VuData->GetChannelGroupMembers(handle, group);
}

void CloseLiveStream(void) 
{ 
  VuData->CloseLiveStream();
};

bool OpenLiveStream(const PVR_CHANNEL &channel) 
{ 
  if (!VuData || !VuData->IsConnected())
    return false;

  return VuData->OpenLiveStream(channel);
}

const char * GetLiveStreamURL(const PVR_CHANNEL &channel) 
{ 
  if (!VuData || !VuData->IsConnected())
    return "";

  return VuData->GetLiveStreamURL(channel);
}

/** UNUSED API FUNCTIONS */
PVR_ERROR SignalStatus(PVR_SIGNAL_STATUS &signalStatus) { return PVR_ERROR_NO_ERROR; }
PVR_ERROR GetStreamProperties(PVR_STREAM_PROPERTIES* pProperties) { return PVR_ERROR_NOT_IMPLEMENTED; } 
void DemuxAbort(void) { return; }
DemuxPacket* DemuxRead(void) { return NULL; }
PVR_ERROR DialogChannelScan(void) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR CallMenuHook(const PVR_MENUHOOK &menuhook) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DeleteChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR RenameChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR MoveChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DialogChannelSettings(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
PVR_ERROR DialogAddChannel(const PVR_CHANNEL &channel) { return PVR_ERROR_NOT_IMPLEMENTED; }
bool OpenRecordedStream(const PVR_RECORDING &recording) { return false; }
void CloseRecordedStream(void) {}
int ReadRecordedStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekRecordedStream(long long iPosition, int iWhence /* = SEEK_SET */) { return 0; }
long long PositionRecordedStream(void) { return -1; }
long long LengthRecordedStream(void) { return 0; }
void DemuxReset(void) {}
void DemuxFlush(void) {}
int ReadLiveStream(unsigned char *pBuffer, unsigned int iBufferSize) { return 0; }
long long SeekLiveStream(long long iPosition, int iWhence /* = SEEK_SET */) { return -1; }
long long PositionLiveStream(void) { return -1; }
long long LengthLiveStream(void) { return -1; }
}
